lib/file_reader.js

```diff
@@ -18,55 +18,62 @@ function checkOS() {
   }
 }
 
-var libraries = [];
-var compiledPackages = false;
-
-function buildParams(package, type) {
-
-  if (package.indexOf('=') > -1) {
-    var singlePackage = package.split('=');
-    var packageName = singlePackage[0];
-    var packageVersion = singlePackage[1];
-    var p = {
-      'name': packageName,
-      'version': packageVersion,
-      'headers': null,
-      'os': null
-    };
-  }
-
-  if (package == '[compiled]' || type === 'compiled') {
-    compiledPackages = true;
-  } else if (p && compiledPackages) {
+function buildParams(p, type) {
+  if (typeof p === 'object' && type === 'compiled') {
     // check OS here
     var osPlatform = checkOS();
     p.os = osPlatform;
+    p.headers = false; 
     packagePath = path.join(p.os, p.name, p.version + '.tar.gz');
-    p['awsPath'] = packagePath;
-    p['src'] = url.resolve(MASON_BUCKET, p.awsPath).replace(/\s/g, '');
-    p['dst'] = path.join(appDir, 'mason_packages', p.awsPath).replace(/\s/g, '').replace('.tar.gz', '');
-    libraries.push(p);
-  } else if (p && package != '[headers]' && !compiledPackages) {
+    p.awsPath = packagePath;
+    p.src = url.resolve(MASON_BUCKET, p.awsPath).replace(/\s/g, '');
+    p.dst = path.join(appDir, 'mason_packages', p.awsPath).replace(/\s/g, '').replace('.tar.gz', '');
+  } else if (typeof p === 'object' && type === 'header') {
     packagePath = path.join('headers', p.name, p.version + '.tar.gz');
-    p['awsPath'] = packagePath;
-    p['headers'] = true;
-    p['src'] = url.resolve(MASON_BUCKET, p.awsPath).replace(/\s/g, '');
-    p['dst'] = path.join(appDir, 'mason_packages', p.awsPath).replace(/\s/g, '').replace('.tar.gz', '');
-    libraries.push(p);
+    p.awsPath = packagePath;
+    p.headers = true;
+    p.os = '';
+    p.src = url.resolve(MASON_BUCKET, p.awsPath).replace(/\s/g, '');
+    p.dst = path.join(appDir, 'mason_packages', p.awsPath).replace(/\s/g, '').replace('.tar.gz', '');
+  }else{
+    p = null;
   }
+  if (p) return p; 
+}
 
-  return libraries;
+function generatePackageObject(p){
+  var singlePackage = p.split('=');
+  var package = {
+    'name': singlePackage[0],
+    'version': singlePackage[1]
+  };
+  return package;
 }
 
 function parseLibraries(fileContents, callback) {
+  var libraries = [];
+  var packageType;
+
   if (fileContents.toString().split('\n')[0] !== '[headers]') {
     var err = new Error();
     err.message = 'Headers must be declared before compiled packages.';
     return callback(err, null);
   }
 
   fileContents.toString().split('\n').forEach(function(line) {
-    buildParams(line);
+    if (line.indexOf('=') > -1) {
+      var p = generatePackageObject(line);
+    }else if (line === '[compiled]'){
+      packageType = 'compiled'; 
+    }else if (line === '[headers]'){
+      packageType = 'header'; 
+    }
+
+    var package = buildParams(p, packageType);
```

Comment:
It looks like `buildParams(p, packageType);` could get called with `p` never being set if `if (line.indexOf('=') > -1) {` never is true. Could this happen?

Reply:
you're right - it's fixed! although I will look into this more in the testing branch since it's unnecessary to call `buildParams` with the declarative `header` or `compiled` strings.